### PR TITLE
Added ability to navigate with menu buttons

### PR DESCRIPTION
**Pull request by DarkMatterMatt**
Added ability to navigate with browser forward/backward buttons as per #48 

- Moved item/module loading from the mod (data set) loader. Now loadData only gets mod data and stores it appropriately. loadModules and loadItems have been added.

- Functions that change the hash now have an optional parameter called dontChangeHash

- Stopped the last item being removed - it causes errors and there's no reason to remove it

Also fixes furnace bug #60

### CODE REVIEW

**Comment by DarkMatterMatt** [comment]
some weird stuff going on - I'll look into it tomorrow. It _usually_ works...

**Comment by DarkMatterMatt** [comment]
**Add load/save/reset settings**
Settings page:
 - Added reset settings button (resets to default)
 - Added save/load custom settings button (stored in localStorage.settings, excludes the mod/data-set)

Save/load buttons are hidden when localStorage is no accessible

loadSettings now loads saved custom settings by default if all settings are default (if the user likes the default settings, then they can save default settings as custom settings so they aren't forced off the default settings or they can simply not save the custom settings)

Fixed bug where hash wasn't updated when browsing to calc.html (with no hash)

Moved hashchange detection initialization out of the DOMContentLoaded listener - it doesn't need to be there.

File: display.js [hunk 1]
```diff
@@ -883,14 +883,16 @@ RecipeTable.prototype = {
...
-function display() {
+function display(dontChangeHash) {
     // Update the display of the target rate text boxes, if needed.
     for (var i = 0; i < build_targets.length; i++) {
         build_targets[i].getRate()
     }
     var totals = globalTotals
 
-    window.location.hash = "#" + formatSettings()
+    if (!dontChangeHash) {
```
Comment (KirkMcDonald) [inline display.js]: I don't like flags which are double-negatives like this. I would prefer a name which is affirmative, like "keepHash".

File: init.js [hunk 2]
```diff
@@ -96,82 +185,29 @@ function loadData(modName, settings) {
...
-                        factory.beaconModule = module
-                        factory.beaconCount = count
-                    }
-                }
-            }
-        }
-        initDone = true
-        itemUpdate()
+        loadModules(settings)
+        loadItems(settings)
```
Comment (KirkMcDonald) [inline init.js]: Why did you change the order of loading the build targets and loading the module spec?

File: init.js [hunk 3]
```diff
@@ -48,6 +58,85 @@ function reset() {
...
+            } else if (type == "r") {
+                target.setRate(parts[2])
+            } else {
+                throw new Error("unknown target type")
+            }
+        }
+    } else if (!build_targets.length) {
+        addTarget()
+    }
+    itemUpdate(dontChangeHash)
```
Comment (KirkMcDonald) [inline init.js]: This should not be called here. It should be called in loadData(), immediately following the loadItems() and loadModules() calls.

File: init.js [hunk 4]
```diff
@@ -48,6 +58,85 @@ function reset() {
...
+            var type = parts[1]
+            if (type == "f") {
+                target.setFactories(parts[2])
+            } else if (type == "r") {
+                target.setRate(parts[2])
+            } else {
+                throw new Error("unknown target type")
+            }
+        }
+    } else if (!build_targets.length) {
```
Comment (KirkMcDonald) [inline init.js]: Why did you add this check?

File: settings.js [hunk 5]
```diff
@@ -141,7 +141,7 @@ function setMinimumAssembler(min) {
 function renderFurnace(settings) {
     var furnace = spec.furnace
     if ("furnace" in settings) {
-        furnace = settings.furnace
+        furnace = {name: settings.furnace}
```
Comment (KirkMcDonald) [inline settings.js]: I don't think this is the right fix. Only the name is used. The variable should be renamed furnaceName, and the name used consistently throughout.

**Comment by KirkMcDonald** [comment]
FYI, I have just pushed 8330172f5f244d9248824a7f586b4c310f95aec8, which directly impacts portions of this PR and will require resolving the merge conflicts.

**Comment by DarkMatterMatt** [comment]
> I don't like that this PR addresses at least three different issues. (Default settings, forward/back buttons, and the furnace setting thing.) This isn't enough to make me reject the PR, but in the future, please keep things more focused.

Sure, I (believe it or not) tried to.
Do you know how to push future commits without them automatically being added here? It's my first time doing this pull request thing.

I'll work on your comments

**Comment by KirkMcDonald** [comment]
I am also not very familiar with how github does PRs, but I believe the PR applies to a specific topic branch in your fork (in this case, the master branch). So you could push changes to other branches, and they should not affect this PR.

**Comment by DarkMatterMatt** [comment]
Closing because I'm splitting browser navigation and setting saving into different pull requests (for personal experience as well as making the pull request less confusing)

**Comment by DarkMatterMatt** [comment]
I don't know how to PM you, so here's my question: should the load/save/reset settings also handle the mod/data-set or should it ignore it? I'm in two minds about it -

On one hand the mod is on the settings page, and everything else is being saved so it makes sense for the mod to be included (and the mod should be loaded by default if the user is using the mod).

On the other hand the mod is unlikely to be changed, and if the user changes it they probably don't want it to be changed back when they load their settings.

**Comment by DarkMatterMatt** [comment]
I'm also thinking that the ignore-list should be automatically stored in localStorage (in a future commit) because it is _very_ unlikely that the user would want individual ignore-lists (and prefer to start with nothing ignored) for each design. The mod/data-set setting could be included with the ignore-list and be automatically processed separate from the other settings.

**Comment by KirkMcDonald** [comment]
I believe that the dataset must be part of the stored settings. In the future, when I add datasets other than vanilla, some of the settings will gain additional options when using those mods. For example, Bob's adds additional assemblers, and so the "minimum assembler" setting may be set to something that is not present in the vanilla dataset.

Indeed, the number of settings might itself change depending on the dataset, but that's something I'll be addressing in the future.

This is also why the settings tab re-renders itself whenever the dataset is changed.

As for the ignore list, I see it as being more akin to the user's selected items and modules than to their settings. The user is designing a factory, and I see the items they wish to ignore as being part of defining that factory, rather than a setting defining how they wish to use the calculator. If that makes sense. This distinction is somewhat vague, I admit, but I'm not convinced that this should be a saved setting.
